### ISSUE TITLE
New Rule: no-default-case

### ISSUE THREAD

**Issue by tupe12334** [eslint/eslint#20091]
### Rule details

Disallow default cases in switch statements to enforce explicit exhaustive handling, with an option to allow empty placeholders for documentation.

### Related ECMAScript feature

https://github.com/tc39/proposal-pattern-matching

### What type of rule is this?

Warns about a potential problem

### Example code

```js
// ❌ This will trigger a warning
function getColorName(color) {
  switch (color) {
    case 'red':
      return 'Red';
    case 'blue':
      return 'Blue';
    default:
      return 'Unknown'; // Warning: default should not be allowed
  }
}
```

## 1. General Applicability (Not TypeScript-Only)

Disallowing `default` in `switch` statements is **not just a TypeScript feature**.  
Many JavaScript codebases use discriminated unions or enum-like objects without TypeScript.

Example:
```js
const STATUS = Object.freeze({
  PENDING: 'pending',
  ACTIVE: 'active',
  INACTIVE: 'inactive',
});

switch (status) {
  case STATUS.PENDING:
    doPending();
    break;
  case STATUS.ACTIVE:
    doActive();
    break;
  case STATUS.INACTIVE:
    doInactive();
    break;
  default:
    doUnknown(); // ❌ Masks missing explicit handling
}
```
This problem exists in plain JavaScript, so it makes sense to address it at the **core level**.

---

## 2. Language-Safety Principle Fits Core's Mission

ESLint core already provides rules aimed at **preventing dangerous patterns**:

- `no-fallthrough`
- `no-console`
- `no-alert`
- `no-eval`

A `default` branch often **silently hides bugs** by catching unexpected values.  
This is the same category of problem: a language feature used in a way that can cause subtle, hard-to-find issues.

> If `no-fallthrough` exists to stop silent behavior between cases, `no-default-case` exists to stop silent behavior for missing cases.

---

## 3. Complements Existing Switch Rules

Core already includes several `switch`-related rules:

- `no-duplicate-case`
- `default-case`
- `default-case-last`

A rule to **forbid `default` entirely** provides a **balanced pair** with `default-case`:

| Rule                | Purpose                        |
|---------------------|--------------------------------|
| `default-case`      | Require a `default` branch     |
| `no-default-case`   | Forbid a `default` branch      |

This symmetry makes ESLint's feature set more complete.

---

## 4. Ecosystem Consistency & Discoverability

Teams should **not need to install niche plugins** for such a fundamental branching policy.

- A core rule provides **one authoritative implementation**.
- Reduces fragmentation across multiple third-party plugins.
- Easier documentation, better support, and more visibility.

This improves **developer experience** for the entire ecosystem.

---

## 5. Future-Proofing with Pattern Matching

The [ECMAScript Pattern Matching Proposal](https://github.com/tc39/proposal-pattern-matching) moves JavaScript toward explicit, exhaustive matching.

Example (future syntax):
```js
match (status) {
  when 'pending': doPending(),
  when 'active': doActive(),
  when 'inactive': doInactive(),
  // ❌ Missing cases would be a runtime error
}
```

A `default` case today **encourages silent fallback behavior**.  
Removing it prepares codebases for modern, safer branching styles.

---

## 6. Clear, Minimal Config Surface

The rule is extremely **simple**:

- **Default:** Disallow `default` entirely.
- **Optional flag:** `allowEmpty` to permit comment-only placeholders.

This is low maintenance and easy to explain, making it a good fit for core.

---

## 7. Good Developer Experience & Autofix Support

ESLint core can provide:

- Reliable, safe autofixes.
- Consistent warning messages.
- Integration with existing tooling.

Example autofix:
```js
// Input
default: 
  console.log('Unexpected');

// Autofix (allowEmpty: true)
default: 
  // Intentionally left blank
```

---

## Counterpoint: “Isn’t This Too Opinionated?”

> *“Some teams **want** default branches. Isn’t this rule too opinionated for core?”*

**Answer:**  
Core already includes **opinionated-but-common rules** like:

- `no-console`
- `prefer-const`
- `eqeqeq`

These rules are optional, but widely used.  
Similarly, `no-default-case` would be **off by default**, and teams can **opt in** if they value explicit handling.

---

## Summary Table

| Reason                               | Why It Fits Core |
|--------------------------------------|------------------|
| General JavaScript applicability     | ✅ |
| Part of language safety and quality  | ✅ |
| Complements existing switch rules    | ✅ |
| Ecosystem consistency and visibility | ✅ |
| Future-proof for pattern matching    | ✅ |
| Simple config and maintenance        | ✅ |

---

**In short:**  
`no-default-case` strengthens core by preventing silent bugs, aligning with modern JS patterns, and giving teams a clear, supported path to explicit, predictable branching.

### Participation

- [x] I am willing to submit a pull request to implement this rule.

**Comment by tupe12334** [comment]
I am currently working on this and have implemented it locally. I will open a pull request soon.
https://eslint.org/docs/latest/contribute/work-on-issue#claiming-an-issue

**Comment by mdjermanovic** [comment]
Hi @tupe12334! Thank for the proposal, but per our [rules policies](https://eslint.org/blog/2020/05/changes-to-rules-policies/), this rule doesn't qualify for inclusion in the core as it isn't related to a new ECMAScript feature. Also, it only disallows syntax, so the same can be enforced by configuring the [`no-restricted-syntax`](https://eslint.org/docs/latest/rules/no-restricted-syntax) rule:

* `SwitchCase[test=null]` selector can be used to entirely disallow default cases: [Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVzdHJpY3RlZC1zeW50YXg6IFtcImVycm9yXCIsXG4gICAge1xuICAgICAgICBzZWxlY3RvcjogXCJTd2l0Y2hDYXNlW3Rlc3Q9bnVsbF1cIixcbiAgICAgICAgbWVzc2FnZTogXCJEb24ndCB1c2UgZGVmYXVsdCBjYXNlLlwiXG4gICAgfVxuXSovXG5cbi8vIGJhZFxuZnVuY3Rpb24gZ2V0Q29sb3JOYW1lKGNvbG9yKSB7XG4gIHN3aXRjaCAoY29sb3IpIHtcbiAgICBjYXNlICdyZWQnOlxuICAgICAgcmV0dXJuICdSZWQnO1xuICAgIGNhc2UgJ2JsdWUnOlxuICAgICAgcmV0dXJuICdCbHVlJztcbiAgICBkZWZhdWx0OlxuICAgICAgcmV0dXJuICdVbmtub3duJzsgLy8gV2FybmluZzogZGVmYXVsdCBzaG91bGQgbm90IGJlIGFsbG93ZWRcbiAgfVxufVxuXG4vLyBiYWRcbmZ1bmN0aW9uIGdldENvbG9yTmFtZTEoY29sb3IpIHtcbiAgc3dpdGNoIChjb2xvcikge1xuICAgIGNhc2UgJ3JlZCc6XG4gICAgICByZXR1cm4gJ1JlZCc7XG4gICAgY2FzZSAnYmx1ZSc6XG4gICAgICByZXR1cm4gJ0JsdWUnO1xuICAgIGRlZmF1bHQ6XG4gICAgICAvLyBJbnRlbnRpb25hbGx5IGxlZnQgYmxhbmtcbiAgfVxufVxuXG4vLyBnb29kXG5mdW5jdGlvbiBnZXRDb2xvck5hbWUyKGNvbG9yKSB7XG4gIHN3aXRjaCAoY29sb3IpIHtcbiAgICBjYXNlICdyZWQnOlxuICAgICAgcmV0dXJuICdSZWQnO1xuICAgIGNhc2UgJ2JsdWUnOlxuICAgICAgcmV0dXJuICdCbHVlJztcbiAgfVxufSIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==)
* `SwitchCase[test=null][consequent.length > 0]` can be used to allow only empty default cases: [Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVzdHJpY3RlZC1zeW50YXg6IFtcImVycm9yXCIsXG4gICAge1xuICAgICAgICBzZWxlY3RvcjogXCJTd2l0Y2hDYXNlW3Rlc3Q9bnVsbF1bY29uc2VxdWVudC5sZW5ndGggPiAwXVwiLFxuICAgICAgICBtZXNzYWdlOiBcIkRvbid0IHVzZSBkZWZhdWx0IGNhc2UuXCJcbiAgICB9XG5dKi9cblxuLy8gYmFkXG5mdW5jdGlvbiBnZXRDb2xvck5hbWUoY29sb3IpIHtcbiAgc3dpdGNoIChjb2xvcikge1xuICAgIGNhc2UgJ3JlZCc6XG4gICAgICByZXR1cm4gJ1JlZCc7XG4gICAgY2FzZSAnYmx1ZSc6XG4gICAgICByZXR1cm4gJ0JsdWUnO1xuICAgIGRlZmF1bHQ6XG4gICAgICByZXR1cm4gJ1Vua25vd24nOyAvLyBXYXJuaW5nOiBkZWZhdWx0IHNob3VsZCBub3QgYmUgYWxsb3dlZFxuICB9XG59XG5cbi8vIGdvb2RcbmZ1bmN0aW9uIGdldENvbG9yTmFtZTEoY29sb3IpIHtcbiAgc3dpdGNoIChjb2xvcikge1xuICAgIGNhc2UgJ3JlZCc6XG4gICAgICByZXR1cm4gJ1JlZCc7XG4gICAgY2FzZSAnYmx1ZSc6XG4gICAgICByZXR1cm4gJ0JsdWUnO1xuICAgIGRlZmF1bHQ6XG4gICAgICAvLyBJbnRlbnRpb25hbGx5IGxlZnQgYmxhbmtcbiAgfVxufVxuXG4vLyBnb29kXG5mdW5jdGlvbiBnZXRDb2xvck5hbWUyKGNvbG9yKSB7XG4gIHN3aXRjaCAoY29sb3IpIHtcbiAgICBjYXNlICdyZWQnOlxuICAgICAgcmV0dXJuICdSZWQnO1xuICAgIGNhc2UgJ2JsdWUnOlxuICAgICAgcmV0dXJuICdCbHVlJztcbiAgfVxufSIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==)
